core/client/fs/src/main/java/alluxio/client/block/stream/BlockInStream.java

```diff
@@ -71,51 +74,72 @@
-    if (Configuration.getBoolean(PropertyKey.USER_SHORT_CIRCUIT_ENABLED)
-        && !NettyUtils.isDomainSocketSupported(address)
-        && blockSource == BlockInStreamSource.LOCAL) {
+  public static BlockInStream create(FileSystemContext context, BlockInfo info,
+      WorkerNetAddress dataSource, BlockInStreamSource dataSourceType, InStreamOptions options)
+      throws IOException {
+    URIStatus status = options.getStatus();
+    OpenFileOptions readOptions = options.getOptions();
+
+    boolean promote = readOptions.getReadType().isPromote();
+
+    long blockId = info.getBlockId();
+    long blockSize = info.getLength();
+
+    // Construct the partial read request
+    Protocol.ReadRequest.Builder builder =
+        Protocol.ReadRequest.newBuilder().setBlockId(blockId).setPromote(promote);
+    if (status.isPersisted()) { // Add UFS fallback options
+      builder.setOpenUfsBlockOptions(options.getOpenUfsBlockOptions(blockId));
+    }
+
+    boolean shortCircuit = Configuration.getBoolean(PropertyKey.USER_SHORT_CIRCUIT_ENABLED);
+    boolean sourceSupportsDomainSocket = !NettyUtils.isDomainSocketSupported(dataSource);
+    boolean sourceIsLocal = dataSourceType == BlockInStreamSource.LOCAL;
+
+    // Short circuit
+    if (sourceIsLocal && shortCircuit && !sourceSupportsDomainSocket) {
+      LOG.debug("Creating short circuit input stream for block {} @ {}", blockId, dataSource);
       try {
-        LOG.debug("Creating short circuit input stream for block {} @ {}", blockId, address);
-        return createLocalBlockInStream(context, address, blockId, blockSize, options);
+        return createLocalBlockInStream(context, dataSource, blockId, blockSize, options);
       } catch (NotFoundException e) {
         // Failed to do short circuit read because the block is not available in Alluxio.
-        // We will try to read from UFS via netty. So this exception is ignored.
-        LOG.warn("Failed to create short circuit input stream for block {} @ {}", blockId, address);
+        // We will try to read from via netty. So this exception is ignored.
+        LOG.warn("Failed to create short circuit input stream for block {} @ {}. Falling back to "
+            + "network transfer", blockId, dataSource);
       }
     }
-    Protocol.ReadRequest.Builder builder = Protocol.ReadRequest.newBuilder().setBlockId(blockId)
-        .setPromote(options.getAlluxioStorageType().isPromote());
-    if (openUfsBlockOptions != null) {
-      builder.setOpenUfsBlockOptions(openUfsBlockOptions);
-    }
 
-    LOG.debug("Creating netty input stream for block {} @ {} from client {}", blockId, address,
-        NetworkAddressUtils.getClientHostName());
-    return createNettyBlockInStream(context, address, blockSource, builder.buildPartial(),
+    // Netty
+    LOG.debug("Creating netty input stream for block {} @ {} from client {} reading through {}",
+        blockId, dataSource, NetworkAddressUtils.getClientHostName(), dataSource);
```

Comment:
NIT: is this `NetworkAddressUtils.getClientHostName()` expected to be expensive in some cases? if yes, let's put the `LOG.debug` inside a `if (LOG.isDebug())` branch

Reply:
This is cached so it should just be a variable access.